lib/pbench/agent/tool_meister_client.py

```diff
@@ -87,149 +83,126 @@ def __init__(
+            )
+        else:
+            self.logger.debug(
+                "constructing SignalExporter() object using host %s:%s," " name: %s",
+                self.redis_host,
+                self.redis_port,
+                self.publisher_name,
             )
+            sig_pub = state_signals.SignalExporter(
+                self.publisher_name,
+                redis_host=self.redis_host,
+                redis_port=self.redis_port,
+            )
+        sig_pub.initialize_and_wait(
+            1,
+            list(tm_allowed_actions),
+            tag="from_pbench_client",
+        )
+        self.sig_pub = sig_pub
+        self.logger.debug("constructed SignalExporter() object")
         return self
 
     def __exit__(self, *args):
-        """On context exit, just close down the to client channel object."""
-        if self.to_client_chan is not None:
-            self.to_client_chan.close()
+        """On context exit, just close down the to SignalExporter object."""
+        self.sig_pub.shutdown()
 
-    def publish(self, group, directory, action, args=None):
-        """publish - marshal a JSON document formed from the group, directory,
-        action, and args arguments.
+    def publish(
+        self,
+        group: str,
+        directory: Union[str, Path, None],
+        action: str,
+        args: Any = None,
+    ) -> int:
+        """Publish a state signal formed from the group, directory, action,
+        and args arguments, using the SignalExporter instance.  It waits for
+        responses from subscribers (TDS, and the TMs indirectly) before
+        continuing.
 
         Returns 0 on success, 1 on failure; logs are also written for any
         errors encountered.
+
         """
-        if action not in api_tm_allowed_actions:
-            return 1
+
         # The published message contains four pieces of information:
         #   {
-        #     "action": "< 'start' | 'stop' | 'send' | 'kill' >",
+        #     "action": "< 'init' | 'start' | 'stop' | 'send' | 'end' | 'terminate' >",
         #     "group": "< the tool group name for the tools to operate on >",
         #     "directory": "< the local directory path to store collected data >"
         #     "args": "< arbitrary argument payload for a particular action >"
         #   }
         # The caller of tool-meister-client must be sure the directory argument
         # is accessible by the Tool Data Sink instance.
-        self.logger.debug("publish %s on chan %s", action, self.from_client_channel)
-        msg = dict(action=action, group=group, directory=str(directory), args=args)
+
+        metadata = dict(
+            group=group,
+            # This check/conversion is necessary, as the pbench scripts
+            # sometimes pass in the directory as a str, and sometimes as
+            # pathlib.PosixPath
+            directory=None if directory is None else str(directory),
```

Comment:
The "check" is presumably necessary because sometimes `directory is None` and we don't want to pass `str(None)`; the "conversion" is because we're passing a "path-like object" which might be `str` or `Path`. (I.e., the comment claims to explain "check/conversion" but actually only explains "conversion".)

Reply:
See if you like https://github.com/Maxusmusti/pbench/pull/13.

Reply:
Yeah, I like Peter's version a bit better, but I'm fine with this having been marked resolved (although we generally prefer to leave that step to the author of the comment).